### PR TITLE
RFC Build API

### PR DESCRIPTION
**Pull request by hayd**
This allows a generic way to build manuals/documentation, as shown in LightGraphs. But improved a bit, namely by having the correctly named markdown files, rather than have markdown in the build.jl. For example
#### docs/linalg.md

```
*LightGraphs.jl* provides the following matrix operations on both directed and
undirected graphs:

## Adjacency

{{adjacency_matrix, adjacency_spectrum}}

## Laplacian

{{
    laplacian_matrix
    laplacian_spectrum
}}
```
#### docs/build.jl

```
using Lexicon
using LightGraphs
# files(["linalg.md"], LightGraphs)  # old
generate_for(LightGraphs, "linalg.md", ...)
```

Documentation is then generated for each file inside `docs/_generated` subdirectory, with output like: https://lightgraphsjl.readthedocs.org/en/latest/linalg/ but with the enumeration trick I mentioned:

screenshot

Note: Hopefully we'll be able to use the same (use ideas from both) formatting when generating api (using Lexicon.save), perhaps grouping by docstring (though that would also make sense here too).

---

TODO
- [x] filter out docstrings defined elsewhere (e.g. reverse) Edit: I _think_ this is working now.
- [ ] group by docstrings (e.g. signatures followed by their shared docstring, repeat)
- [x] get source links working (not sure how this is already done) Edit: this is a hack, copy and pasting from elsewhere in Lexicon.
- [x] The function name `files` sucks, better suggestions welcome. :) Edit: maybe `generate_for(LightGraphs, "linalg.md", ...)` Edit: am going with that for now...

cc @MichaelHatherly @sbromberger @peter1000 Thoughts?

### CODE REVIEW

File: src/build.jl [hunk 1]
```diff
@@ -0,0 +1,105 @@
+# This file generated the Markdown documentation files.
+
+# The @file macro generates the documentation for a particular file
+# where the {{method1, methods2}} includes the documentation for each method
+# via the `buildwriter` function.
+
+# Currently this prints the methodtable followed by the docstring.
+
+using LightGraphs
```
Comment (hayd) [inline src/build.jl]: Ah, this causes the build failure. I think this is something silly... I think I need to do

    using $m

inside the right scope (src/build.jl).

**Comment by hayd** [comment]
I don't understand these `@doc` errors, as it works for me locally (passes tests and generates docs). The using/eval/toplevel trick may be 0.4 only however.

File: src/build.jl [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+
+buildwriter(ex::Expr, m::Module) = :(print(file, $(esc(ex))))
+
+buildwriter(t::AbstractString, m::Module) = Expr(:block,
+    [buildwriter(p, iseven(n), m) for (n, p) in enumerate(split(t, r"{{|}}"))]...
+)
+
+buildwriter(part, isdef, m) = isdef ?
+    begin
+        parts = Expr(:vect, [:(($(parse(p))), Base.Docs.@doc($(parse(p)))) for p in filter(x -> strip(x) != "", split(part, r"\s*[, \n]\s*"))]...)
```
Comment (hayd) [inline src/build.jl]: It's unclear why I added this (the Base.Docs), it appeared to be necessary to fix locally, but it's obviously not what we want! Presumably there's a better way!

File: src/build.jl [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+# The @file macro generates the documentation for a particular file
+# where the {{method1, methods2}} includes the documentation for each method
+# via the `buildwriter` function.
+
+# Currently this prints the methodtable followed by the docstring.
+
+macro file(fn, md_string, m::Module) buildfile(fn, md_string, m) end
+
+function generate_for(m::Module, file_names...; doc_dir="docs", gen_dir="_generated")
+    eval(Expr(:toplevel, quote using $(symbol(m)) end))
```
Comment (MichaelHatherly) [inline src/build.jl]: In 0.3 ``using`` can't be put directly into quotes. I believe it was a fixed quite recently. You'd need to used ``Expr(:using, ...)`` instead until we've moved to 0.4/0.5.

**Comment by MichaelHatherly** [comment]
> I don't understand these ``@doc`` errors, as it works for me locally (passes tests and generates docs). The using/eval/toplevel trick may be 0.4 only however.

See inline comment.

Thanks @hayd. I'll have a proper look at this on the weekend.

I think if we're going to have a functional interface rather than direct macro calls for this then you should be able to eliminate the ``@file``/``eval`` combo and probably just have ``eval`` somewhere in there. Maybe, I've only skimmed over the code though.

File: src/build.jl (hunk 3 above)
Comment (hayd) [inline src/build.jl]: Unfortunately I don't think that works, as it needs to be done on the top level (IIUC using is a syntax error when not done globally). ?

**Comment by hayd** [comment]
I removed the `@file` export so `generate_for` is the only export for build. Internally it still uses `@file` but I imagine this could be refactored out in future. :)

File: src/build.jl (hunk 3 above)
Comment (MichaelHatherly) [inline src/build.jl]: ``eval(Expr(:toplevel, Expr(:using, :Docile)))``? That works I think.

**Comment by MichaelHatherly** [comment]
> I removed the ``@file`` export so ``generate_for`` is the only export for build. Internally it still uses ``@file`` but I imagine this could be refactored out in future. :)

Sure, that sounds fine.

File: src/build.jl (hunk 3 above)
Comment (hayd) [inline src/build.jl]: Ah ha, that works. Groovy.

Comment (MichaelHatherly) [inline src/build.jl]: Is it necessary to import the module ``m`` like this? Shouldn't it already be imported if you're passing around a reference to it?

Comment (hayd) [inline src/build.jl]: Yes, I think this is because of macros (defining them on the Lexicon side).

**Comment by hayd** [comment]
Tests are passing now (though I haven't written any facts for this, and I'm not sure how tbh). Source links etc. are working. To do the grouping a slightly different approach might be needed (haven't tried to do so yet). But I think this is already pretty neat.

Check out along with the LightGraph branch: https://github.com/hayd/LightGraphs.jl/tree/new_doc

**Comment by hayd** [comment]
Another thought on this would be:

    {{
        > foo
    }}

which would become (assuming foo() == 23):

```
julia> foo()
23
```

**Comment by hayd** [comment]
Note: ATM this does not work for non-exported functions. That may or may not be desired... come to think about it (being specific, Lexicon.foo, for non-exported methods seems like a good idea doc/worthy).

**Comment by hayd** [comment]
I have an implementation that also supports

```
{{
    >! foo() = 23
    > foo()
    > bar()
    > a = [1];
    > a[3]
}}
```
becomes

    ```
    julia> foo()
    23
    julia> bar()
    UndefVarError: bar not defined
    julia> a = [1];
    julia> a[3]
    BoundsError: attempt to access 1-element Array{Int64,1}:
     1
      at index [3]
    ```

which is pretty cool. The code's messy but it's pretty cool I think.

**Comment by MichaelHatherly** [comment]
I like the REPL idea! Would be really nice to generalise it a bit so that we can have user-defined ``{{...}}`` blocks kind of like the metamacros in Docile that can do arbitrary textual transformations.

The use of ``eval`` is still a bit annoying, but I think I've got a way to sort that out. I'll write up a quick gist of what I'm thinking.

**Comment by hayd** [comment]
I'm of the opinion, pick a good API then clean up (the evals) later :)

metamacros sound interesting, but I think they can be added later... I probably need to do better to catch "syntax" errors, currently:

1. is it all julia commands (> or >!)
2. otherwise assume they are separated functions

Really we could just enforce the newline separation for functions. i.e. not allow

```
{{func1, func2 fun3}
```
Maybe that'd allow additional syntaxes to be added later...

**Comment by MichaelHatherly** [comment]
Here's the gist: https://gist.github.com/MichaelHatherly/c944d4a204de74f30e30

> but I think they can be added later

With the "directives" in the gist, the REPL could be added with something like:

```jl
function parsebrackets(:: directive"repl", text)
    # parse, eval, do magic things
    # return the text to splice back into it's place
end
```

and would then be used with

```
{{#repl#
    > a = 1 + 2
}}
```

the ``#repl#`` syntax is up debate though. I'd prefer something explicit and named rather than searching for specific syntax within a block.

> I'm of the opinion, pick a good API then clean up (the evals) later :)

If we can avoid the ``eval`` right now, then we should.

**Comment by hayd** [comment]
I'd be good with:
```
{{repl
    > a = 1 + 2
}}
```
That's very similar to Markdown code-language syntax, so it feels like "honest" markdown.

**Comment by hayd** [comment]
That directive stuff is *really* nice.

My repl code is in this [commit](https://github.com/hayd/Lexicon.jl/commit/5f3a6ca84673509cd0328aa7af6c622102d01e39), I've no idea whether you could do it without eval.

So the final syntax would be:

```
{{
   func1
   func2
}}
```
and
```
{{repl
    >! foo() = 23
    > foo()
}}
```

I think that takes us a long way, and the majority of uses. Though it's great that users could make their own directives. :)

**Comment by MichaelHatherly** [comment]
A problem with not having any marker to denote the directive is that you could potentially run into the following:

```md
Docs for the REPL function:

{{repl}}
```

What happens here? Is it an empty REPL directive or the docs for ``repl``? I'm still simpathetic to having the syntax be as honest as possible though. Perhaps just a ``:`` after the directive name:

```md
{{repl:
    julia> a = 1 + 2
}}
```

```md
{{repl}}
```

*should* probably avoid the ambiguity.

For the ``>!`` syntax, perhaps just having another directive called ``define`` would be better:

```md
{{define:
    foo() = 23
}}

{{repl:
    julia> foo()
}}
```

``define`` wouldn't splice anything back into the text, only execute the code found inside. Is that the behaviour you were going for with ``>!``?

**Comment by MichaelHatherly** [comment]
> I've no idea whether you could do it without eval.

I think it could be. You'd just need to return something kind of ``REPLResults`` object that stored the unparsed and parsed form of the text:

```jl
function Lexicon.parsebrackets(:: directive"repl", text)
    unparsed_lines = ...
    parsedlines = ...
    quote
        REPLResult($(unparsed_lines), $(parsedlines))
    end
end
```

and define a  ``writer`` method than knows what to do with it to display properly:

```jl
function Lexicon.writer(io :: IO, replresults :: REPLResults)
    ...
end
```

Haven't tried that yet, so not 100% sure.

**Comment by hayd** [comment]
That's a good question, but I think we should just say that

```
{{repl}}
```
was the same as
```
{{
    repl
}}
```
in the same way that ```` ```jl``` ```` is resolved.

---

Yes, the >! was hidden, IIRC python documentation has some similar functionality (so for example you can do imports). and the >! syntax is often used to mean "spoiler". :s 

The question there is, is it frequent you want to do that (and mix them together)... probably not!

---

Very excited about this!!!

**Comment by MichaelHatherly** [comment]
> but I think we should just say that

If that can be done without causing any ambiguities and too much additional complexity then that's fine.

To go to the the absolute edge case: what about directives that don't take an arguments/text, ie.

```md
{{clear_defines}}
```

Not sure how common that kind of thing would be in practise though. Requiring a ``:`` for it would make it a bit less magical I think.

---

The ``>!`` syntax was a bit confusing at first. Having the intention spelt out and separate from displayed code would be a lot clearer I think.

---

> Very excited about this!!!

Feel free to extract whatever you need from that gist and add it to this PR, though perhaps a PR against the ``lexicon-next`` branch would be better given that this is beginning to evolve quite a bit.

**Comment by hayd** [comment]
> To go to the the absolute edge case: what about directives that don't take an arguments/text, i

I don't follow you.

```
{{clear_defines}}
```
would not be a directive, it's be the same as
```
{{
    clear_defines
}}
```

Similarly in commonmark, the dialect is taken:

    ```jl
    foo()
    ```

if there's spaces then the top line is used for the class: [Example](http://johnmacfarlane.net/babelmark2/?text=%60%60%60jl+bar%0Afoo()%0A%60%60%60).

---

Your code looks **a lot** cleaner, so if there is some way to meld the two/use yours that'd be fantastic. 

Do we need to support 0.3 ? (for *generating* documentation I don't think it's so important).

Will see if I have some time over the weekend. :s 

---

Do you have any thoughts on using `generate_for` (functionality) for the API sections. i.e. create {{}}-syntax markdown, and then `@file` it.

**Comment by MichaelHatherly** [comment]
> would not be a directive

That's the trouble though. If you want it to be a directive then there doesn't seem to be a way to do that. I'm quite liking the

```md
{{repl:
...
}}
```

syntax. One thing that is totally different between code blocks in markdown and these ``{{...}}`` blocks is that the one just keeps the text verbatim whereas the other does some arbitrarily complex transformation. That's probably enough of a difference to warrant different syntax.

> Do we need to support 0.3

No. ``using Compat`` was there just out of habit. You can happily ignore the existence of 0.3 for this if you'd like.

> Do you have any thoughts on using ``generate_for``

We should be able to build this directly on top of ``loadfile``. Just something like the following pseudo-code:

```jl
function makedocs(files :: Pair...)
    for (k, v) in files
        loadfile(k, v)
    end
end

function makedocs(dir)
    for file in readdir(dir)
        loadfile(file, joinpath("build", file))
    end
end
```

with the name ``makedocs`` just to add another name to the mix...

File: src/build.jl [hunk 4]
```diff
@@ -0,0 +1,324 @@
+
+"""The `@file` macro takes filename and a `md_string` and a module m.
+
+First it expands the {{}} syntax with respect to the module, i.e. docstings
```
Comment (peter1000) [inline src/build.jl]: docstings > docstrings

**Comment by hayd** [comment]
An empty directive could be:
```
{{clear_defines
}}
```
Perhaps we could compromise and have the : should be optional i.e. 

```
{{clear_defines:}}
```
e.g. with

```
const DIRECTIVE_MARKER = r"^(\w+)\s*[:\n](.*)"
# or even more forgiving
const DIRECTIVE_MARKER = r"^(\w[^\n]*)[:\n](.*)"
```
then default if match is nothing or match[2] is "".

---

You have cleaned this up a lot... let me see if I can change that! :)

**Comment by MichaelHatherly** [comment]
> Perhaps we could compromise and have the : should be optional i.e. 

Sounds fine to me.

**Comment by MichaelHatherly** [comment]
I realised that we could probably just unify the directives and metamacros so that the same syntax could be used in docstrings as well as external files. A start of that potential change is in the branch https://github.com/MichaelHatherly/Docile.jl/tree/rewrite. Lexicon will keep ``@file`` and ``loadfile`` which will just call out to ``Docile.Directives.build`` for parsing ``{{...}}`` syntax.

**Comment by MichaelHatherly** [comment]
OK, I've also gone and simplified the ``rewrite`` branch in Lexicon. So any PRs should be against either the Docile or Lexicon ``rewrite`` branches. I'll move over bits that are useful from the other branches as and when they are needed.

**Comment by peter1000** [comment]
I noticed you seem to have changed your style adding extra spaces arounde the double colons?
`function loadfile(path :: AbstractString, out :: AbstractString)`

Any particular reason?

**Comment by MichaelHatherly** [comment]
Atom does some odd highlighting when you don't have spaces, which I don't particularly like... I've not got any attachment to the style though so I should probably just stop doing it.

**Comment by peter1000** [comment]
no problem: after I saw you posting at  the *atom-julia-client* I also tried atom and I'm very pleased.  so far.Thanks :+1: 

I don't think the spaces are a problem except that each definition line will become again longer.

Anyway, I will still wait for a final julia 0.4 to look into documentation related thing.

**Comment by hayd** [comment]
About the rewrite branch: What's the Docile.Directives.build business? Do I need some special version of Docile. Or should I use directives from your gist, I'm a little confused what the intention is (is the gist newer?).

I'm kinda sullking that this branch is "working nicely" (but a hack)... :p

**Comment by MichaelHatherly** [comment]
> About the rewrite branch: What's the Docile.Directives.build business? Do I need some special version of Docile.

The ``rewrite`` branch in Docile is also needed. So further developments will, for the moment, be done against ``Lexicon/rewrite`` and ``Docile/rewrite``. Once things are all working with both of those, then they can be merged into ``master``, probably after Julia 0.4 is released (or at least in RC).

> Or should I use directives from your gist, I'm a little confused what the intention is (is the gist newer?).

The ``Docile/rewrite`` branch now contains the basic directives, since I realised we can also use them in docstrings rather than having separate ``{{foobar:...}}`` and ``!!foobar(...)`` syntax that basically do the same things as each other.

> I'm kinda sullking that this branch is "working nicely" (but a hack)... :p

Just about everything in this PR (and the doctest PR) is still very much relevant (:+1:) for the ``rewrite`` branches:

- doctesting multiple submodules
- "repl" directive, which just needs to be defined as a ``parsebracket(:: directive"repl", text)``
- better signatures and source links

---

As an example of the changes in the rewrite branches, we can do the following once the REPL directive is implemented:

```jl
module M

using Docile.Docs

# apply directives parsing to each docstring prior to storing them
addhook(directives)

"""
{{repl:
> a = 1
> b = 2
> c = a + 2b
}}
"""
foobar(x) = ...

end
```

It turns out that it's [surprisingly straightforward](https://github.com/MichaelHatherly/Docile.jl/blob/cfc6151b923db9655ce58ade433233a8a071984a/src/Docs.jl#L34-L50) to hook into the docsystem in 0.4 and allow for just about any kind of transformation to take place.

**Comment by hayd** [comment]
Some work in the direction, hope it helps: https://github.com/hayd/Lexicon.jl/tree/curly

At least one eval slipped in there (and not resolved/tried to resolve the using Module aspect) ...

**Comment by MichaelHatherly** [comment]
@hayd, I've left a couple of comments inline in https://github.com/hayd/Lexicon.jl/commit/a84e60b48c8ec13ac5218097fc959acff60300b6.

**Comment by MichaelHatherly** [comment]
With all the work on Documenter I'll close this as stale.

**Pull request closed without merging**